### ISSUE TITLE
remote_foreground() doesn't work

### ISSUE THREAD

**Issue by ronaaron** [vim/vim#11641]
### Steps to reproduce

1. Start gvim and note the server name (echo v:servername)
2. Start another gvim and call remote(foreground('thatservername'))
3. Note that the first gvim does **not** move to the foreground.

### Expected behaviour

Expect that the first gvim moves to the foreground. It does not, though **wmctrl** can be shelled from the second gvim to have the desired effect.

### Version of Vim

9.0.975 GTK3 Huge build

### Environment

Linux Mint 21, Cinnamon desktop

**Comment by brammool** [comment]
I tried a few things, but it looks like the foreground() function just stopped working.
Both with Motif and with GTK.  I suspect that the X11 implementation has some mechanism
that MS-Windows has been using for a longer time, that prevents applications to put themselves
into the foreground even when there is no user interaction.
Calling foreground() in the server does work, it's only when the function call is invoked remotely that it doesn't appear to work.

**Comment by ronaaron** [comment]
Right, I made a typo in the call. Just "foreground()" also doesn't work over the server connection.

I don't think your supposition about user-interaction is correct, since I can exec a wmctrl from the second gvim and cause the first one to become foreground. In fact, that's what I've been doing for a while (I don't know how long, but at least a year or two).

**Comment by ronaaron** [comment]
In 8th I'm using SDL2, so I did a test and ran "SDL_SetWindowAlwaysOnTop()" after a 5 second sleep where in the meantime I brought other applications to the foreground. The affected 8th window became 'topmost' after 5 seconds and was visible above all others. Don't know if that helps, but it might point in a direction.

**Comment by ronaaron** [comment]
I don't know why gtk_window_present() doesn't work, but I did see a note that it doesn't work on OSX (not of interest at the moment).

This replacement for gui_mch_foreground() seems to work, though I'm unsure if it's the best way to go about it:

```
    void
gui_mch_set_foreground(void)
{
    gtk_window_set_keep_above(GTK_WINDOW(gui.mainwin), (gboolean) TRUE);
    do_sleep(10,0);
    gtk_window_set_keep_above(GTK_WINDOW(gui.mainwin), (gboolean) FALSE);
}

```

**Comment by brammool** [comment]
Ron Aaron wrote:

> I don't know why gtk_window_present() doesn't work, but I did see a
> note that it doesn't work on OSX (not of interest at the moment).
> 
> This replacement for gui_mch_foreground() seems to work, though I'm
> unsure if it's the best way to go about it:
> 
> ```
>     void
> gui_mch_set_foreground(void)
> {
>     gtk_window_set_keep_above(GTK_WINDOW(gui.mainwin), (gboolean) TRUE);
>     do_sleep(10,0);
>     gtk_window_set_keep_above(GTK_WINDOW(gui.mainwin), (gboolean) FALSE);
> }
> 
> ```

For me this briefly shows the window on top, but then it's below again.

I can make it work if there is a call to gtk_window_present() in
between.  Let me send a patch with that and see if this causes a problem
somehow.

-- 
Some of the well known MS-Windows errors:
	EMEMORY		Memory error caused by..., eh...
	ELICENSE	Your license has expired, give us more money!
	EMOUSE		Mouse moved, reinstall Windows
	EILLEGAL	Illegal error, you are not allowed to see this
	EVIRUS		Undetectable virus found

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ronaaron** [comment]
Yes, the patch does work. Thanks!

**Comment by ronaaron** [comment]
Oof. Well, it's not working 100% of the time. I don't yet understand what the problem is though,.

**Comment by ronaaron** [comment]
What I found was that in my vimrc I was doing "packload" which, since I'm using some heavy packages, apparently interfered with the server stuff sometimes. I moved it to the end of a script which gets autoloaded rather late, and it seems to work much better.

While I'm on the topic: would it be possible to have some way to set the server name in vimscript?
